File: R/Tutorial 4.R
### Preliminaries ###

setwd("")

### Loading Libraries ###
library(ISLR)
library(tidyverse)
library(caret)
library(rsample)
library(FNN)
# glmnet library for regularization with generalized linear models.
library(glmnet)

### Regularization ###

## Regression ##

# Loading a dataset
df = data.frame(Auto)
head(df)
?Auto

# We will regress mpg on all predictors.

# No regularization
lm_model = lm(mpg ~ . - name,data=df)
summary(lm_model)

# Matrices
?matrix
A = matrix(data = c(1, 2, 3, 4), nrow=2, ncol=2, byrow = TRUE)
A
B = matrix(data = c(1, 2, 3, 4), nrow=2, ncol=2, byrow = FALSE)
B
A.T = t(A)
A.T
A%*%A
B-A.T
B+A.T
solve(A)
A%*%solve(A)
A*A
det(A)
diag(A)

# Ridge regularization
# Defining X and y matrices
X = as.matrix(df[,c('cylinders','displacement','horsepower','weight','acceleration','year','origin')])
X
y = as.matrix(df[,'mpg'])

# Note. We do not need to manually scale variables when using glmnet().

# Running ridge regression
ridge_reg = glmnet(x=X,y=y, family='gaussian', alpha=0)
plot(ridge_reg)
print(ridge_reg)
coef(ridge_reg,s=0.01)

# Cross-validation to choose lambda
ridge_reg_cv = cv.glmnet(x=X,y=y,family='gaussian',alpha=0,nfolds = 10)
plot(ridge_reg_cv)
ridge_reg_cv = cv.glmnet(x=X,y=y,family='gaussian',alpha=0,lambda.min=0.000001)
plot(ridge_reg_cv)
lmin = ridge_reg_cv$lambda.min
l1se = ridge_reg_cv$lambda.1se
coef(ridge_reg_cv,s=lmin)
coef(ridge_reg_cv,s=l1se)

# LASSO regularization
lasso_reg = glmnet(X,y,family='gaussian',alpha=1)
plot(lasso_reg)
coef(lasso_reg,s=0.01)

lasso_reg_cv = cv.glmnet(X,y,family='gaussian',alpha=1,lambda.min=0.000001)
plot(lasso_reg_cv)
lmin = ridge_reg_cv$lambda.min
l1se = ridge_reg_cv$lambda.1se
coef(lasso_reg_cv,s=lmin)
coef(lasso_reg_cv,s=l1se)
lasso_reg_cv$glmnet.fit

# Elastic Net
net_reg_cv = cv.glmnet(X,y,family='gaussian',alpha=0.5,lambda.min=0.000001)
plot(net_reg_cv)

# Regularization using caret
testgrid = expand.grid(alpha=seq(0,1,0.1),lambda=seq(0.000001,7,0.001))

train_control = trainControl(method='cv',number = 10)
net_caret = train(mpg ~ . - name, data=df, method='glmnet', trControl = train_control, tuneGrid = expand.grid(alpha = 1, lambda=lmin))
net_caret

# Running Regularized Regression with CV for both alpha and lambda
net_caret = train(mpg ~ . - name, data=df, method='glmnet', trControl = train_control, tuneGrid = testgrid)
net_caret
net_caret = train(mpg ~ . - name, data=df, method='glmnet', trControl = train_control, tuneLength = 11)
net_caret
net_caret$finalModel
coef(net_caret$finalModel,s=net_caret$bestTune$lambda)

## Classification ##
df = data.frame(Smarket)
?Smarket

df_split = initial_split(data = df, prop = 0.7)
df_train = training(df_split)
df_test = testing(df_split)

X_train = model.matrix(Direction ~ factor(Year) + . - 1 - Today, df_train)
y_train = as.matrix(df_train[,'Direction'])
X_test = model.matrix(Direction ~ factor(Year) + . - 1 - Today, df_test)
y_test = as.matrix(df_test[,'Direction'])
# Using glmnet() directly
# Ridge penalty
ridge_logit_cv = cv.glmnet(x=X_train,y=as.factor(y_train), type.measure='class',family='binomial', alpha=0, lambda.min=0.0000001)
plot(ridge_logit_cv)
ridge_logit_cv$lambda.min
# LASSO penalty
lasso_logit_cv = cv.glmnet(x=X_train,y=as.factor(y_train), type.measure='class',family='binomial', alpha=1, lambda.min=0.0000001)
plot(lasso_logit_cv)
lasso_logit_cv$lambda.min

# Using caret
train_control = trainControl(method='cv',number = 10)
net_class = train(factor(Direction) ~ factor(Year) + . - Today, data=df_train, method='glmnet', trControl = train_control, tuneLength = 50)
net_class
net_pred = predict(net_class,df_test)
net_pred
confusionMatrix(net_pred,as.factor(y_test))

# Sensitivity to grid
net_class = train(factor(Direction) ~ factor(Year) + . - Today, data=df_train, method='glmnet', trControl = train_control, tuneGrid = testgrid)
net_class
net_pred = predict(net_class,df_test)
net_pred
confusionMatrix(net_pred,as.factor(y_test))

# Checking k=5
train_control = trainControl(method='cv',number = 5)
net_class = train(factor(Direction) ~ factor(Year) + . - Today, data=df_train, method='glmnet', trControl = train_control, tuneLength = 50)
net_class
net_pred = predict(net_class,df_test)
net_pred
confusionMatrix(net_pred,as.factor(y_test))

File: Python/Tutorial 4.py
### Preliminaries ###
# Loading os library to interact with system
import os
# Set the working directory
os.chdir("")

### Loading Libraries ###
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegressionCV
from sklearn.model_selection import train_test_split
# Sklearn functions to implement Ridge
from sklearn.linear_model import Ridge
from sklearn.linear_model import RidgeCV
# Sklearn functions to implement elastic net
from sklearn.linear_model import ElasticNet
from sklearn.linear_model import ElasticNetCV
# Sklearn functions to plot regularization path
from sklearn.linear_model import enet_path

### Regularization ###

## Regression ##

# Loading a dataset
df = pd.read_csv("../Datasets/Auto.csv",index_col=0)

# We will regress mpg on all predictors.
df_X = df.drop(['mpg','name'],axis=1)
df_y = df['mpg']
lm_model = sm.OLS(exog=df_X,endog=df_y).fit()
lm_model.summary()
# There are a few variables that have insignificant coefficients. In addition, recall from the pairs
# plot we have looked at a few times that many of these explanatory variables are highly
# correlated. So, this model may benefit from using regularization to reduce the influence
# of unneeded explanatory variables that likely lead to overfitting.

# Ridge regularization

# At the moment, statsmodels' implementation of regularization does not store information 
# about the "regularization path", or the resulting models for different values of lambda.
# Most importantly, it does not implement CV. Thus, we will use sklearn for regularization.
# sklearn's implementation (and statsmodels') is based on glmnet() in R.

## Note on sklearn regularization ##
# Unlike glmnet(), sklearn does not automatically scale data when using regularization.
# Remember to appropriately scale your data before you estimate any model. For the sake of
# brevity, we will not do this in the code here. You should try to scale the data yourself
# and see whether the results change.

# A note on terminology. The parameter l1_ratio corresponds to alpha in the glmnet R package 
# while alpha corresponds to the lambda parameter in glmnet.

# Ridge() only estimates one model for a specific value of lambda. In principle, we can
# also estimate ridge regression with ElasticNet(), but the documentation states it is
# unstable for l1_ratio<=0.01.
ridge_reg = Ridge(alpha=1).fit(X=df_X,y=df_y)
ridge_reg.coef_

# To get the regularization path, we need to use enet_path(). The function outputs
# the values of lambda that it tests, the corresponding coefficients, and convergence
# information (duality gap, this is not important). This also does not work with l1_ratio
# <=0.01, so for the sake of the graph we will use 0.02.
alphas, ridge_path, _ = enet_path(X=df_X,y=df_y,l1_ratio=0.02,n_alphas=100,eps=1e-9)

# We can plot the path as follows.
for coef in ridge_path:
    plt.semilogx(alphas, coef)
plt.xlabel("alpha")
plt.ylabel("coefficients")
plt.title("Ridge Regularization Path")
plt.axis("tight")

# To implement cross-validation to choose lambda, we can use ElasticNetCV().
ridge_reg_cv = ElasticNetCV(l1_ratio=0.02,cv=10,n_alphas=1000,eps=1e-9).fit(X=df_X,y=df_y)
# Here, eps controls the range of alphas that are tested. It is defined as
# eps = alpha_min/alpha_max.
# We can get the optimal lambda using .alpha_, and the corresponding coefficients by
# using .coef_.
ridge_reg_cv.alpha_
ridge_reg_cv.coef_
# To plot the cross-validation curve, we can access the alphas tested using .alphas_
# and the mse's using .mse_path_. .mse_path_ returns an array of size(n_alphas,n_folds),
# meaning we need to average the mse across each fold manually.
alphas = ridge_reg_cv.alphas_
mses = np.mean(ridge_reg_cv.mse_path_,axis=1)
plt.semilogx(alphas,mses)
plt.xlabel("Log(alpha)")
plt.ylabel("MSE")
plt.title("MSE Curve")
plt.axis("tight")

# LASSO is very similar. We will implement it using ElasticNet().
lasso_reg = ElasticNet(alpha=1,l1_ratio=1).fit(X=df_X,y=df_y)
lasso_reg.coef_

# Regularization path with LASSO
alphas, lasso_path, _ = enet_path(X=df_X,y=df_y,l1_ratio=1,n_alphas=100,eps=1e-9)

# We can plot the path as follows.
for coef in lasso_path:
    plt.semilogx(alphas, coef)
plt.xlabel("alpha")
plt.ylabel("coefficients")
plt.title("LASSO Regularization Path")
plt.axis("tight")

# Cross-validation
lasso_reg_cv = ElasticNetCV(l1_ratio=1,cv=10,n_alphas=1000,eps=1e-9).fit(X=df_X,y=df_y)
lasso_reg_cv.alpha_
lasso_reg_cv.coef_
# MSE Curve
alphas = lasso_reg_cv.alphas_
mses = np.mean(lasso_reg_cv.mse_path_,axis=1)
plt.semilogx(alphas,mses)
plt.xlabel("Log(alpha)")
plt.ylabel("MSE")
plt.title("MSE Curve")
plt.axis("tight")

# With both LASSO and Ridge, the optimal amount of shrinkage is miniscule. This is probably
# because in this example the coefficients are already quite small, even for the
# insignificant coefficients.

# ElasticNetCV() can also be used to choose l1_ratio via cross-validation by passing in
# a list of values to test for l1_ratio. The documentation recommends the following, but
# evenly spaced values between 0 and 1 are also sensisble.
l1_list = [.1, .5, .7, .9, .95, .99, 1]
net_cv = ElasticNetCV(l1_ratio=l1_list,cv=10,n_alphas=1000,eps=1e-9).fit(X=df_X,y=df_y)
net_cv.alpha_
net_cv.l1_ratio_
# Note that LASSO is chosen via CV, but the optimal lambda is still tiny. Thus, not much
# regularization is applied.
net_cv.coef_

## Classification ##
# Loading dataset
df = pd.read_csv("../Datasets/Smarket.csv",index_col=0)
# Data splitting
df_train, df_test = train_test_split(df,test_size=0.3)
# The explanatory variables are dummies for Year, and all other variables except for Today.
# Dummies can easily be generated by pandas' get_dummies() function.
year_ind = pd.get_dummies(df_train.Year)
year_ind
# They are encoded as True/False but will automatically get converted to 1/0 as needed in
# estimation.
# Note that the column names have type int.
year_ind.columns[0].dtype
# This is because get_dummies uses the possible values for the categorical variable as
# each column name, which in this case happened to be an integer. This will be a problem
# for some estimation algorithms, including logit below. So, we will convert the column
# names to strings.
year_ind.columns = year_ind.columns.astype(str)
year_ind.columns[0]
X_train = pd.concat((year_ind,df_train.drop(['Year','Today','Direction'],axis=1)),axis=1)
# Here we used pd.concat to join the columns of the dummies dataframe and the dataframe 
# of the remaining explanatory variables.
y_train = df_train.loc[:,'Direction']
testyear_ind = pd.get_dummies(df_test.Year)
testyear_ind.columns = testyear_ind.columns.astype(str)
X_test = pd.concat((testyear_ind,df_test.drop(['Year','Today','Direction'],axis=1)),axis=1)
y_test = df_test.loc[:,'Direction']

# Classification via logistic regression with regularization is similar to OLS. Recall
# that regularization is directly built into the functionality of LogisticRegression().
# Here, we will directly jump to using CV via the LogisticRegressionCV() function. Instead
# of specifying lambda, LogisticRegression() and LogisticRegressionCV() specify C, defined
# as the inverse of lambda.
ridge_logit_cv = LogisticRegressionCV(Cs=1000,cv=10,penalty='l2').fit(X=X_train,y=y_train)
ridge_logit_cv.C_
# The model output stores the regularization path. Let's plot the results as before.
# MSE Curve
lambdas = 1/ridge_logit_cv.Cs_
# The model output stores the classification accuracy for each class in .scores_ as a key in
# a dictionary. The value for each key is an array of accuracy scores and has size (n_folds, n_Cs).
# In this binary classification case, there is only one key for the first class that appears in
# the data, which in this case was 'Up'. We will have to average the accuracy across all folds.
acc = np.mean(ridge_logit_cv.scores_['Up'],axis=0)
plt.semilogx(lambdas,acc)
plt.xlabel("Log(Lambda)")
plt.ylabel("Accuracy")
plt.title("Accuracy Curve")
plt.axis("tight")
# LASSO would be identically applied, just change penalty to 'l1'. We cannot use the default
# bfgs solver however.
lasso_logit_cv = LogisticRegressionCV(Cs=1000,cv=10,penalty='l1',solver='liblinear').fit(X=X_train,y=y_train)
lasso_logit_cv.C_
lambdas = 1/lasso_logit_cv.Cs_
acc = np.mean(lasso_logit_cv.scores_['Up'],axis=0)
plt.semilogx(lambdas,acc)
plt.xlabel("Log(Lambda)")
plt.ylabel("Accuracy")
plt.title("Accuracy Curve")
plt.axis("tight")
# To estimate elastic net and choose the l1_ratio via CV, change penalty to 'elastic_net',
# and pass in a list of values for l1_ratio. As discussed in the R tutorial, regularized
# results can be sensitive to the number of folds, and the grid you use for l1_ratio, so
# be sure to check multiple values.